go/vt/vtgate/executor_dml_test.go

```diff
@@ -388,6 +388,202 @@ func TestInsertSharded(t *testing.T) {
+		BindVariables: map[string]*querypb.BindVariable{
+			"music_id": sqltypes.Int64BindVariable(2),
+		},
+	}, {
+		Sql: "select user_id from music_user_map where music_id = :music_id",
+		BindVariables: map[string]*querypb.BindVariable{
+			"music_id": sqltypes.Int64BindVariable(3),
+		},
+	}, {
+		Sql: "select user_id from music_user_map where music_id = :music_id",
+		BindVariables: map[string]*querypb.BindVariable{
+			"music_id": sqltypes.Int64BindVariable(4),
+		},
+	}, {
+		Sql: "select user_id from music_user_map where music_id = :music_id",
+		BindVariables: map[string]*querypb.BindVariable{
+			"music_id": sqltypes.Int64BindVariable(5),
+		},
+	}, {
+		Sql: "select user_id from music_user_map where music_id = :music_id",
+		BindVariables: map[string]*querypb.BindVariable{
+			"music_id": sqltypes.Int64BindVariable(6),
+		},
+	}, {
+		Sql: "insert ignore into ins_lookup(fromcol, tocol) values (:fromcol0, :tocol0), (:fromcol1, :tocol1), (:fromcol2, :tocol2), (:fromcol3, :tocol3), (:fromcol4, :tocol4)",
+		BindVariables: map[string]*querypb.BindVariable{
+			"fromcol0": sqltypes.Int64BindVariable(1),
+			"tocol0":   sqltypes.Uint64BindVariable(1),
+			"fromcol1": sqltypes.Int64BindVariable(3),
+			"tocol1":   sqltypes.Uint64BindVariable(1),
+			"fromcol2": sqltypes.Int64BindVariable(4),
+			"tocol2":   sqltypes.Uint64BindVariable(1),
+			"fromcol3": sqltypes.Int64BindVariable(5),
+			"tocol3":   sqltypes.Uint64BindVariable(1),
+			"fromcol4": sqltypes.Int64BindVariable(6),
+			"tocol4":   sqltypes.Uint64BindVariable(3),
+		},
+	}, {
+		Sql: "select fromcol from ins_lookup where fromcol = :fromcol and tocol = :tocol",
+		BindVariables: map[string]*querypb.BindVariable{
+			"fromcol": sqltypes.Int64BindVariable(1),
+			"tocol":   sqltypes.Uint64BindVariable(1),
+		},
+	}, {
+		Sql: "select fromcol from ins_lookup where fromcol = :fromcol and tocol = :tocol",
+		BindVariables: map[string]*querypb.BindVariable{
+			"fromcol": sqltypes.Int64BindVariable(3),
+			"tocol":   sqltypes.Uint64BindVariable(1),
+		},
+	}, {
+		Sql: "select fromcol from ins_lookup where fromcol = :fromcol and tocol = :tocol",
+		BindVariables: map[string]*querypb.BindVariable{
+			"fromcol": sqltypes.Int64BindVariable(4),
+			"tocol":   sqltypes.Uint64BindVariable(1),
+		},
+	}, {
+		Sql: "select fromcol from ins_lookup where fromcol = :fromcol and tocol = :tocol",
+		BindVariables: map[string]*querypb.BindVariable{
+			"fromcol": sqltypes.Int64BindVariable(5),
+			"tocol":   sqltypes.Uint64BindVariable(1),
+		},
+	}, {
+		Sql: "select fromcol from ins_lookup where fromcol = :fromcol and tocol = :tocol",
+		BindVariables: map[string]*querypb.BindVariable{
+			"fromcol": sqltypes.Int64BindVariable(6),
+			"tocol":   sqltypes.Uint64BindVariable(3),
+		},
+	}}
+	if !reflect.DeepEqual(sbclookup.Queries, wantQueries) {
+		t.Errorf("sbclookup.Queries: \n%+v, want \n%+v", sbclookup.Queries, wantQueries)
+	}
+
+	// Test the 0 rows case,
```

Comment:
We discussed a bit about this offline. The outcome of the discussion was that the common use cases will be handled correctly by the current behavior. The part that was not obvious was that VTGate would do the right thing if all the keyspace ids matched. If they didn't match, the likelihood of it being an app bug was very high. So, the safe thing to do would be to not forward the statement in such cases.